File: modules/expo-face-landmarker/ios/ExpoFaceLandmarkerModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import CoreImage
import UIKit

// MARK: - Result types

struct Blendshape: Codable {
  let name: String
  let score: Float
}

struct LandmarkPoint: Codable {
  let x: Float
  let y: Float
  let z: Float
}

// MARK: - Module

public class ExpoFaceLandmarkerModule: Module {

  private var faceLandmarker: FaceLandmarker?

  public func definition() -> ModuleDefinition {
    Name("ExpoFaceLandmarker")

    OnCreate {
      self.setupFaceLandmarker()
    }

    // Returns { blendshapes: [{name, score}], landmarks: [{x, y, z}] } | null
    AsyncFunction("detectFaces") { (base64Jpeg: String, promise: Promise) in
      guard let data = Data(base64Encoded: base64Jpeg),
            let uiImage = UIImage(data: data),
            let cgImage = uiImage.cgImage else {
        promise.resolve(nil)
        return
      }

      guard let landmarker = self.faceLandmarker else {
        promise.reject("LANDMARKER_NOT_READY", "Face landmarker not initialized")
        return
      }

      let mpImage = try MPImage(uiImage: uiImage)
      let result = try landmarker.detect(image: mpImage)

      guard let firstFace = result.faceLandmarks.first else {
        promise.resolve(nil)
        return
      }

      // ── 52 Blendshape coefficients ──
      var blendshapes: [[String: Any]] = []
      if let firstBlendshapes = result.faceBlendshapes.first {
        blendshapes = firstBlendshapes.categories.map { cat in
          ["name": cat.categoryName ?? "unknown", "score": cat.score]
        }
      }

      // ── 478 3D face landmarks (normalized 0-1 screen coords) ──
      let landmarks = firstFace.map { lm in
        ["x": lm.x, "y": lm.y, "z": lm.z]
      }

      promise.resolve([
        "blendshapes": blendshapes,
        "landmarks": landmarks,
      ])
    }

    // Synchronous frame detection (called from Vision Camera frame processor worklet via runOnJS)
    Function("detectFacesSync") { (base64Jpeg: String) -> [String: Any]? in
      guard let data = Data(base64Encoded: base64Jpeg),
            let uiImage = UIImage(data: data),
            let landmarker = self.faceLandmarker else {
        return nil
      }

      do {
        let mpImage = try MPImage(uiImage: uiImage)
        let result = try landmarker.detect(image: mpImage)

        guard let firstFace = result.faceLandmarks.first else { return nil }

        var blendshapes: [[String: Any]] = []
        if let firstBlendshapes = result.faceBlendshapes.first {
          blendshapes = firstBlendshapes.categories.map { cat in
            ["name": cat.categoryName ?? "", "score": cat.score]
          }
        }

        let landmarks = firstFace.map { lm in
          ["x": lm.x, "y": lm.y, "z": lm.z]
        }

        return ["blendshapes": blendshapes, "landmarks": landmarks]
      } catch {
        print("[ExpoFaceLandmarker] Detection error: \(error)")
        return nil
      }
    }
  }

  // MARK: - Private

  private func setupFaceLandmarker() {
    guard let modelPath = Bundle.main.path(forResource: "face_landmarker", ofType: "task") else {
      // Fall back to module bundle
      guard let modulePath = Bundle(for: ExpoFaceLandmarkerModule.self)
              .path(forResource: "face_landmarker", ofType: "task") else {
        print("[ExpoFaceLandmarker] ⚠️ face_landmarker.task model not found in bundle")
        return
      }
      createLandmarker(modelPath: modulePath)
      return
    }
    createLandmarker(modelPath: modelPath)
  }

  private func createLandmarker(modelPath: String) {
    do {
      let options = FaceLandmarkerOptions()
      options.baseOptions.modelAssetPath = modelPath
      options.runningMode = .image
      options.numFaces = 1
      options.outputFaceBlendshapes = true
      options.outputFacialTransformationMatrixes = false
      options.minFaceDetectionConfidence = 0.5
      options.minFacePresenceConfidence = 0.5
      options.minTrackingConfidence = 0.5

      faceLandmarker = try FaceLandmarker(options: options)
      print("[ExpoFaceLandmarker] ✅ FaceLandmarker initialized")
    } catch {
      print("[ExpoFaceLandmarker] ❌ Failed to create FaceLandmarker: \(error)")
    }
  }
}

File: modules/expo-face-landmarker/android/src/main/java/expo/modules/facelandmarker/ExpoFaceLandmarkerModule.kt
package expo.modules.facelandmarker

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker.FaceLandmarkerOptions
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.mrousavy.camera.frameprocessors.FrameProcessorPluginRegistry
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoFaceLandmarkerModule : Module() {

  companion object {
    init {
      // Register our VisionCamera FrameProcessorPlugin so it can be called from JS
      // via VisionCameraProxy.initFrameProcessorPlugin('detectFacesFromFrame')
      FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectFacesFromFrame") { proxy, options ->
        ExpoFaceLandmarkerPlugin(proxy, options)
      }
    }
  }


  private var faceLandmarker: FaceLandmarker? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoFaceLandmarker")

    // Async version for base64 JPEG input (used for one-off calls / iOS)
    AsyncFunction("detectFaces") { base64Jpeg: String, promise: Promise ->
      try {
        val result = runDetection(base64Jpeg)
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("DETECTION_ERROR", e.message, e)
      }
    }

    // Sync version for iOS frame processor use (called via runOnJS from worklet)
    Function("detectFacesSync") { base64Jpeg: String ->
      try {
        runDetection(base64Jpeg)
      } catch (e: Exception) {
        null
      }
    }

    // Android frame processor path: accepts raw RGBA bytes from frame.toArrayBuffer()
    // when pixelFormat="rgb" is set on the VisionCamera Camera component.
    // Called asynchronously from the JS thread after the worklet dispatches the buffer.
    AsyncFunction("detectFacesFromRgba") { width: Int, height: Int, rgbaBytes: ByteArray ->
      try {
        val bitmap = createBitmapFromRgba(width, height, rgbaBytes)
        runDetectionOnBitmap(bitmap)
      } catch (e: Exception) {
        android.util.Log.e("ExpoFaceLandmarker", "❌ detectFacesFromRgba failed: ${e.message}")
        null
      }
    }
  }

  // MARK: - Private

  private fun getLandmarker(): FaceLandmarker? {
    if (faceLandmarker != null) return faceLandmarker
    val context = appContext.reactContext ?: return null
    try {
      val baseOptions = BaseOptions.builder()
        .setModelAssetPath("face_landmarker.task")
        .build()

      val options = FaceLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setRunningMode(RunningMode.IMAGE)
        .setNumFaces(1)
        .setOutputFaceBlendshapes(true)
        .setMinFaceDetectionConfidence(0.5f)
        .setMinFacePresenceConfidence(0.5f)
        .setMinTrackingConfidence(0.5f)
        .build()

      faceLandmarker = FaceLandmarker.createFromOptions(context, options)
      android.util.Log.d("ExpoFaceLandmarker", "✅ FaceLandmarker initialized")
    } catch (e: Exception) {
      android.util.Log.e("ExpoFaceLandmarker", "❌ Failed to create FaceLandmarker: ${e.message}")
    }
    return faceLandmarker
  }

  private fun runDetection(base64Jpeg: String): Map<String, Any>? {
    val bytes = Base64.decode(base64Jpeg, Base64.DEFAULT)
    val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size) ?: return null
    return runDetectionOnBitmap(bitmap)
  }

  /**
   * Converts raw RGBA bytes (from VisionCamera's frame.toArrayBuffer() with pixelFormat="rgb")
   * into an Android Bitmap suitable for MediaPipe.
   *
   * VisionCamera's "rgb" pixelFormat on Android produces RGBA_8888 bytes.
   * Each pixel is 4 bytes in [R, G, B, A] order, which we pack into Android's
   * ARGB_8888 int format as 0xAARRGGBB.
   */
  private fun createBitmapFromRgba(width: Int, height: Int, rgbaBytes: ByteArray): Bitmap {
    val intPixels = IntArray(width * height)
    for (i in intPixels.indices) {
      val r = rgbaBytes[i * 4].toInt() and 0xFF
      val g = rgbaBytes[i * 4 + 1].toInt() and 0xFF
      val b = rgbaBytes[i * 4 + 2].toInt() and 0xFF
      val a = rgbaBytes[i * 4 + 3].toInt() and 0xFF
      intPixels[i] = (a shl 24) or (r shl 16) or (g shl 8) or b
    }
    return Bitmap.createBitmap(intPixels, width, height, Bitmap.Config.ARGB_8888)
  }

  private fun runDetectionOnBitmap(bitmap: Bitmap): Map<String, Any>? {
    val landmarker = getLandmarker() ?: return null

    val mpImage = BitmapImageBuilder(bitmap).build()
    val result = landmarker.detect(mpImage)

    val faceLandmarksList = result.faceLandmarks()
    if (faceLandmarksList.isEmpty()) return null

    val firstFace = faceLandmarksList[0]

    // ── 478 3D landmarks ──
    val landmarks: List<Map<String, Float>> = firstFace.map { lm ->
      mapOf("x" to lm.x(), "y" to lm.y(), "z" to lm.z())
    }

    // ── 52 Blendshape coefficients ──
    val blendshapes: List<Map<String, Any>> = if (result.faceBlendshapes().isPresent) {
      val firstFaceBlendshapes = result.faceBlendshapes().get()
      if (firstFaceBlendshapes.isNotEmpty()) {
        firstFaceBlendshapes[0].map { cat ->
          mapOf("name" to (cat.categoryName() ?: ""), "score" to cat.score())
        }
      } else emptyList()
    } else emptyList()

    return mapOf(
      "blendshapes" to blendshapes,
      "landmarks" to landmarks,
    )
  }
}
